File: lab5/java/comp26120/dp_kp.java
package comp26120;

import java.util.ArrayList;

public class dp_kp extends KnapSack {
    public dp_kp(String filename) {
	super(filename);
    }

    public void DP(ArrayList<Boolean> solution) {
	ArrayList<Integer> v = item_values;
	ArrayList<Integer> wv = item_weights;
	int n = Nitems;
	int W = Capacity;
	
	// the dynamic programming function for the knapsack problem
	// the code was adapted from p17 of http://www.es.ele.tue.nl/education/5MC10/solutions/knapsack.pdf

	// v array holds the values / profits / benefits of the items
	// wv array holds the sizes / weights of the items
	// n is the total number of items
	// W is the constraint (the weight capacity of the knapsack)
	// solution: a 1 in position n means pack item number n+1. A zero means do not pack it.

	ArrayList<ArrayList<Integer>> V = new ArrayList<ArrayList<Integer>>(n + 1);
	ArrayList<ArrayList<Boolean>> keep = new ArrayList<ArrayList<Boolean>>(n + 1);;// 2d arrays for use in the dynamic programming solution
	// keep[][] and V[][] are both of size (n+1)*(W+1)

	int i, w, K;

	// Initialise V and keep with null objects
	/* ADD CODE HERE */
 
	//  set the values of the zeroth row of the partial solutions table to zero
	/* ADD CODE HERE */


	// main dynamic programming loops , adding one item at a time and looping through weights from 0 to W
	/* ADD CODE HERE */


	// now discover which items were in the optimal solution
	/* ADD CODE HERE */


	
    }
    
    public static void main(String[] args) {
	ArrayList<Boolean> solution;

	dp_kp knapsack = new dp_kp(args[0]);
	solution = new ArrayList<Boolean>(knapsack.Nitems+1);
	solution.add(null); // C implementation has null first object

	knapsack.DP(solution);
	knapsack.check_evaluate_and_print_sol(solution);
    }
}

File: lab5/python/dp_kp.py
import sys
import time

from knapsack import knapsack


class dp(knapsack):
    def __init__(self, filename):
        knapsack.__init__(self, filename)
        self.time = 0

    def DP(self, solution):
        start_time = time.time()
        # Renaming things to keep track of them wrt. names used in algorithm
        v = self.item_values
        wv = self.item_weights
        n = self.Nitems
        W = self.Capacity

        # the dynamic programming function for the knapsack problem
        # the code was adapted from p17 of http://www.es.ele.tue.nl/education/5MC10/solutions/knapsack.pdf

        # v array holds the values / profits / benefits of the items
        # wv array holds the sizes / weights of the items
        # n is the total number of items
        # W is the constraint (the weight capacity of the knapsack)
        # solution: True in position n means pack item number n+1. False means do not pack it.

        # V and Keep should be 2d arrays for use in the dynamic programming solution
        # They are both of size (n + 1)*(W + 1)

        # Initialise V and keep
        V = [[0 for x in range(W + 1)] for y in range(n + 1)]
        Keep = [[False for x in range(W + 1)] for y in range(n + 1)]
        # main dynamic programming loops, adding on item at a time and looping through weights from 0 to W
        for i in range(1, n + 1):
            for j in range(0, W + 1):
                if (wv[i] <= j) and (v[i] + V[i - 1][j - wv[i]] > V[i - 1][j]):
                    V[i][j] = v[i] + V[i - 1][j - wv[i]]
                    Keep[i][j] = True
                else:
                    V[i][j] = V[i - 1][j]
        K = W
        # now discover which items were in the optimal solution
        for i in range(n, 0, -1):
            if Keep[i][K]:
                print("Item %d is in the knapsack" % i)
                solution[i] = True
                K = K - wv[i]
        end_time = time.time()
        self.time = end_time - start_time
        return V[n][W]


knapsk = dp(sys.argv[1])
solution = [False] * (knapsk.Nitems + 1)
knapsk.DP(solution)
knapsk.check_evaluate_and_print_sol(solution)
